### PR TITLE
Instant Send Autolocks

### PR DESCRIPTION
**Pull request by tomasz-ludek**
This pull request implements the improvements described in stories:

- DMD-886 Instant Send / Autolocks - Autolocks Enabled (>4 inputs)
- DMD-887 Instant Send / Autolocks - Autolocks Enabled (<=4 inputs)
- DMD-888 Instant Send / Autolocks - Exceeds Confirmed Balance
- DMD-889 Instant Send / Autolocks - Autolocks are Disabled

It basically adds support for InstantSend Auto Locks.

The new flow works as follows:
If IS and Auto Locks are supported by the network, the wallet will try to send all the transactions using IS Auto Lock, if the transaction doesn't meet the conditions of being a simple transaction and can't be auto locked, but user has enough funds to send it as a standard IS, she will be asked whether wants to add an additional fee and force IS, or rather send it as a regular transaction. If she doesn't have enough funds to send it as a standard IS, it will be automatically send as regular transaction.

It covers the following designs:

send_lessthan4inputs

send_morethan4inputs

send_regular

### CODE REVIEW

**Comment by HashEngineering** [comment]
I will review this after the dashj elements are reviewed.

**Comment by HashEngineering** [comment]
This works as advertised in the screen shots above.  This PR still uses the old "InstantSend is awaiting verification" and "InstantSend verified" text as seen in this screen shot.

screenshot_1547179128

File: wallet/src/de/schildbach/wallet/ui/WalletTransactionsFragment.java [hunk 1]
```diff
@@ -147,6 +148,15 @@ public void onChange(final boolean selfChange) {
         }
     };
 
+    private InstantSendEventListener instantSendEventListener = new InstantSendEventListener() {
```
Comment (HashEngineering) [inline wallet/src/de/schildbach/wallet/ui/WalletTransactionsFragment.java]: This listener is redundant.  

The wallet has a listener for transaction confidence changes and that triggers this listener at line 563:
```
    private final ThrottlingWalletChangeListener transactionChangeListener = new ThrottlingWalletChangeListener(
            THROTTLE_MS) {
        @Override
        public void onThrottledWalletChanged() {
            adapter.notifyDataSetChanged();
        }
    };
```

**Comment by HashEngineering** [comment]
I didn't get any transactions to fail being locked.  

Also, there is no difference between mainnet (autolocks off) and testnet (autolocks) on in how the app behaves, which means that this PR must be released after autolocks are activated.

There have been cases in the past when InstantSend was disabled by the Core team.  We should have a way to handle that.  Your send code does look if InstantSend and Autolocks are activated, so that is a good start.  Will the text at the bottom "InstantSend enabled by default at Zero extra fee" be absent if InstantSend is off?

File: wallet/src/de/schildbach/wallet/ui/WalletTransactionsFragment.java (hunk 1 above)
Comment (tomasz-ludek) [reply wallet/src/de/schildbach/wallet/ui/WalletTransactionsFragment.java]: I removed the InstantSendEventListener from DashJ: [4ccd1bd](https://github.com/dashevo/dashj/commit/4ccd1bd0342c7c24dbd956124bc91fe7cab2e064)
and from the wallet: [0bedf73](https://github.com/HashEngineering/dash-wallet/pull/156/commits/0bedf7318149afb7a7d49210540f7039abb6e914#diff-2be9434186730f60270a0e6862aa8f3bL41)
The `transactionChangeListener` is registred for Transaction Confidence Events and after applying
your fix of transaction confidence changes listener the InstantSendEventListener will be needless.

Also used `IXType.IX_LOCK_FAILED` to verify the status of simple transactions: [8576794](https://github.com/HashEngineering/dash-wallet/commit/8576794d88c246fb08843bdc461242384e4c08d2)

**Comment by tomasz-ludek** [comment]
> There have been cases in the past when InstantSend was disabled by the Core team. We should have a way to handle that. Your send code does look if InstantSend and Autolocks are activated, so that is a good start. Will the text at the bottom "InstantSend enabled by default at Zero extra fee" be absent if InstantSend is off?

It wasn't handled, I've just added a basic support for inactive SPORK_16_INSTANTSEND_AUTOLOCKS: [0bedf73](https://github.com/HashEngineering/dash-wallet/pull/156/commits/0bedf7318149afb7a7d49210540f7039abb6e914)

File: wallet/src/de/schildbach/wallet/ui/WalletActivity.java [hunk 2]
```diff
@@ -186,7 +186,7 @@ protected void onCreate(final Bundle savedInstanceState) {
             upgradeWalletKeyChains(Constants.BIP44_PATH, false);
         }
 
-        if (InstantSend.canAutoLock() && !config.getFastestNetworkAnncmntShown()) {
+        if (!config.getFastestNetworkAnncmntShown()) {
```
[the remark this answers is not in the document]
Comment (tomasz-ludek) [reply wallet/src/de/schildbach/wallet/ui/WalletActivity.java]: The issue with this check is that `InstantSend.canAutoLock()` return *false* in this place until the spork list is synced. My other idea is to use SporkUpdatedEventListener, but to do this we will have to merge https://github.com/dashevo/dashj/pull/13 first.

[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/WalletActivity.java]: That makes sense.

[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/WalletActivity.java]: I do want that PR 13 merged.

[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/WalletActivity.java]: faf48e1 appears to have resolved this.

File: wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java [hunk 3]
```diff
@@ -1294,36 +1299,126 @@ public void run() {
...
+            case REGULAR_PAYMENT: {
+                sendRequest.coinSelector = ZeroConfCoinSelector.get();
+                sendRequest.useInstantSend = false;
+                sendRequest.feePerKb = fees.get(feeCategory);
+                break;
+            }
+        }
+
+        sendRequest.ensureMinRequiredFee = sendRequest.useInstantSend;
+        sendRequest.signInputs = signInputs;
```
[the remark this answers is not in the document]
Comment (tomasz-ludek) [reply wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java]: Shouldn't that be handled by DashJ itself? I mean, the described change will cause throwing the DustySendRequested exception:
https://github.com/dashevo/dashj/blob/dash-0.13/core/src/main/java/org/bitcoinj/wallet/Wallet.java#L4025 (we already catch it in SendCoinsFragment).

[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java]: https://github.com/HashEngineering/dash-wallet/pull/177

When ensureMinRequiredFee is false dust checks are disabled, but the above PR will fix that.  We have to check it ourselves.

[the remark this answers is not in the document]
Comment (tomasz-ludek) [reply wallet/src/de/schildbach/wallet/ui/send/SendCoinsFragment.java]: This issue will be handled within https://github.com/HashEngineering/dash-wallet/pull/177

File: wallet/res/values/strings-extra.xml [hunk 4]
```diff
@@ -1,7 +1,11 @@
 <?xml version='1.0' encoding='UTF-8'?>
 <resources>
 
-	<string name="send_coins_fragment_instantx_enable">Use InstantSend</string>
+    <string name="send_coins_auto_lock_feasible">This transaction will settle instantly at no extra fee.</string>
+	<string name="send_coins_auto_lock_not_feasible">This transaction may take several minutes to settle. Complete instantly for an extra %1$s%2$s?</string>
+	<string name="send_coins_instant_send_not_feasible">This transaction may take several minutes to settle.</string>
+	<string name="send_coins_instant_send_enabled_by_default">InstantSend has been enabled by default at\u00A0Zero\u00A0Extra\u00A0Fee</string>
```
Comment (HashEngineering) [inline wallet/res/values/strings-extra.xml]: this says at Zero extra fee.  We have three different ways of saying this.

Comment (HashEngineering) [reply wallet/res/values/strings-extra.xml]: with no extra fee
at no extra fee
at Zero extra fee

**Comment by HashEngineering** [comment]
This has been merged into a different branch.

File: wallet/res/values/strings-extra.xml [hunk 5]
```diff
@@ -1,7 +1,11 @@
 <?xml version='1.0' encoding='UTF-8'?>
 <resources>
 
-	<string name="send_coins_fragment_instantx_enable">Use InstantSend</string>
+    <string name="send_coins_auto_lock_feasible">This transaction will settle instantly at no extra fee.</string>
```
[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/res/values/strings-extra.xml]: I fixed this.

**Comment by HashEngineering** [comment]
This was merged with this commit:  https://github.com/HashEngineering/dash-wallet/commit/78927ea861db677b46d5327675a6494d5ee11d06

**Pull request closed without merging**